src/util/locking/WrappedExpiringResourceLocker.ts

```diff
@@ -24,14 +23,62 @@ export class WrappedExpiringResourceLocker implements ExpiringResourceLocker {
     this.expiration = expiration;
   }
 
+  public async withReadLock<T>(identifier: ResourceIdentifier, func: (renew: () => void) => T | Promise<T>):
+  Promise<T> {
+    return this.locker.withReadLock(identifier, async(): Promise<T> => this.timePromise(identifier, func));
+  }
+
+  public async withWriteLock<T>(identifier: ResourceIdentifier, func: (renew: () => void) => T | Promise<T>):
+  Promise<T> {
+    return this.locker.withWriteLock(identifier, async(): Promise<T> => this.timePromise(identifier, func));
+  }
+
+  /**
+   * Creates a Promise that either resolves the given input function or rejects if time runs out,
+   * whichever happens first. The input function can reset the timer by calling the `renew` function
+   * it receives. The ResourceIdentifier is only used for logging.
+   */
+  private async timePromise<T>(identifier: ResourceIdentifier, func: (renew: () => void) => T | Promise<T>):
+  Promise<T> {
+    let timer: Timeout;
+    let timerReject: (error: Error) => void;
```

Comment:
```suggestion
    let onTimeout: (error: Error) => void;
```

Reply:
After going through all these comments: it might have been easier to have a small string mapping as a single comment 😄